### ISSUE TITLE
Unified connection & transaction management

### ISSUE THREAD

**Issue by anarazel** [citusdata/citus#704]
Right now there's various pieces of the system that handle connections.
## There's several distinct "classes" of connection handling discernible to me:
1. Router Executor uses one-connection per remote node, keyed by host:port. These connections are established for the lifetime of a session (or until failure). There never is more than one connection. 
2. Real-Time/Task-Track Executor use establish several connections to workers. There is no caching/pooling. On errors there's a number of places where such connections can leak until the end of session.
3. Various sites (e.g. COPY,  multi_shard_transaction.c users) have localized connection management of shorter lived connections, often with one connection per placement.
4.  Various sites use one-off connection management. E.g. everything using ExecuteRemoteQuery().
## There's several forms of transaction duration management:
1. Some sites (e.g. real-time/task-tracker executor) don't care about transactions at all, they just establish new connections. I.e. they emulate read-committed behaviour, from outside of transaction blocks.  That prevents such connections being used after a modification have already been made inside the current transaction, as read-your-own-writes would otherwise be violated.
2. Some sites integrate into postgres' transaction management using transaction callbacks (e.g. router executor after modifications, multi_shard_transaction.c users). Transactions are committed / rolled back when the containing transaction commits. 
3. Other sites use transactions on remote nodes, but commit as soon as the local operation is finished, regardless of the local transaction state (e.g. COPY). That makes local rollbacks confusing, but has the advantage of allowing later reads to return something meaningful, even if a different form of transaction management is used.
## There's several forms of error handling in combination with transactions:
1. No error handling whatsoever.
2. Interrupts are disabled, thus remote commands are completed, even if there's a local cancel and such
3. Remote transactions without 2PC. That allows msater transactions being aborted before the final commit, but not halfway through the remote nodes
4. Use 2PC. Currently that allows for the highest consistency, but we have error recovery (i.e. old 2pc that need to either be committed or rolled back on all nodes) issues.
## That leads to a number of issues:
1. It's hard to understand, explain behaviour
2. code complexity. Lots of duplication, interrupt handling, pg_try/catch blocks, ...
3. When writing new code one has to choose from various choices with conflicting advantages, and has to think about a complicated mix of interactions
4. Interactions between the different modes makes it hard to combine actions using different connection or transaction management.
5. Performance issues due to frequent connection establishment
6. Error due to too many connections (e.g. violating max_connections) / connection establishment (e.g. running out of tcp ports)
## I think we need to unify code and behaviour around this. I think there's several desirable traits:
1. Integrates with postgres transaction scope. I.e. remote commits happen only when local commits occur. Only one transaction hook should exist.
2. Allows for read-your-own-writes, _or_, when that's not viable, reliably produces an error.
3. Users of the connection/transaction management code shouldn't have to deal with transaction, nor error handling. I.e. on error connections will be rolled back. Non-persistent connections are disconnected upon error. 
4. Some state is maintained per connection: 1) in-transaction or not 2) "dirty" 3) exclusive locks acquired 4) cached-or-not
5. Allows, in future, for snapshots of existing connections to be copied into new connections, to allow for real-time executor usage after modifications

Any other desirable features? Not necessarily to be implemented now, but to shape the API. 

Remaining Steps:

- [x]  Central commit handler & Connection lifetime management
- [x] RemoteTransaction handling
- [x] Placement->Connection tracking
- [x] Shard -> Placement tracking to invalidate failed shards and/or error out
- [x] Placement->Connection mapping should be colocation aware (#1098)
- [x] Convert master_modify_multiple_shards() to new API.
- [x] Migrate DDL to new framework
- [x] Move router executor over to new connection & transaction framework.
- [x] Move INSERT .. SELECT, DDL code, master_modify_multiple_shards()  in router executor over to new connection & transaction framework.
- [x] Migration COPY to new framework. (#1096)
- [ ] Convert remaining callers (#1101)
- [x] Remove obsoleted transaction infrastructure.

**Comment by anarazel** [comment]
FWIW, I think the error leading to closing non-cached connections after errors should be tied to resource owners, not transactions.

**Comment by anarazel** [comment]
There's also the question whether we consider 2PC recovery part of this.

**Comment by marcocitus** [comment]
- Connections have a well-defined scope / ownership (e.g. worker-specific DDL, shard-specific DDL, metadata)
- Enables DDL parallelism (>1 connection)
- Allows connections using different roles, mainly current user and super user
- Nice-to-have: We can add connections/workers to an ongoing 2PC

As an API, I quite like the [worker_transaction](https://github.com/citusdata/citus-mx/blob/masterless/main/src/backend/distributed/transaction/worker_transaction.c
) functions in MX where the only thing the caller has to do is call SendCommandToWorkersInParallel  that starts a (recoverable) 2PC across all the workers if one is not already active. This made it fairly straight-forward to add new features. 

SendCommandToWorkers opens a connection to all workers as super-user, which is required to make metadata changes on Citus Cloud. For 2PC, the function reuses the [commit_protocol](https://github.com/citusdata/citus/blob/master/src/backend/distributed/transaction/commit_protocol.c) API that COPY uses through the [multi_shard_transaction](https://github.com/citusdata/citus/blob/master/src/backend/distributed/transaction/multi_shard_transaction.c) functions. We're currently using the multi_shard_transaction functions for DDL and master_modify_multiple_shards as well, but it is less suitable for that since it opens too many connections. worker_transaction and multi_shard_transaction are compatible in the sense that they can be treated as one big 2PC in which all connections work on different, unrelated objects.

It was also fairly straight-forward to [make dropping shards part of the 2PC](https://github.com/citusdata/citus-mx/commit/96d8af4164bba662f438ca4857af3bd8fe01e89a#diff-91ca00ec89d001590223e89aa6c160c5R302) using SendCommandToWorkersInParallel, which allowed rolling back DROP Table and master_apply_delete_command. This is currently not in the main branch because there were some conflicts with 5.2 transactions.

Part of the worker_transaction and multi_shard_transaction API's is to wrap the PGconn in a [TransactionConnection](https://github.com/citusdata/citus/blob/master/src/include/distributed/commit_protocol.h#L48) that contains the necessary state to perform a 2PC.

**Comment by aamederen** [comment]
With @onderkalaci, we have thought about the potential problems of joint use of per-worker transactions and per-placement transactions. The following scenario might be worth addressing in the current context:

1. Open per-worker transactions and send a command X.
2. Open per-placement transactions and send a command Y.

Ignoring shard-pruning, one transaction from (1) and some transactions of (2) will be on the same worker. If command X and command Y both locks the same resource, then we will have a deadlock.

Can this be a problem in the future? 

Currently, in MX, we have a joint use of per-worker transactions and per-placement transactions in DDL propagation logic but have seen no deadlock there yet 😃

**Comment by marcocitus** [comment]
Meeting notes from the meeting on Friday between Andres, Jason and myself. Added notes from some further studying into Citus commands.

General consensus is that we need one place where connections are managed, including opening/caching/closing connection, commit/abort handlers, and tracking the state necessary to guarantee consistency, atomicity, and deadlock avoidance

**Requirements of command types**

We plan to keep the session-level connection cache that is currently used for router SELECT and DML with at most one connection per worker. Using a cached connection connection is essential to low latency, particularly when SSL is involved, while caching more than one connection would create a lot of overhead.

DDL currently opens connections to all shard placement, but this does not scale to very large shard counts or append-distributed tables. Instead, we should open up a small number of connection to allow commands like CREATE INDEX to be performed in parallel, but without running out of connections for append-distributed tables or creating an I/O bottleneck. 

The real-time executor currently opens new connections for each task for parallelism and closes it at the end of the statement. We could potentially reuse connections within the real-time executor which we currently don't. We may also be able to reuse connections that were already open for writes, but this may get complicated (see further down).

COPY on hash-distributed tables always needs as many connections as there are shard placements, since we need to keep a COPY block open for each shard placement in order to be able to send each incoming row to the appropriate shard placement. 

**Connection pool modes**

There are certain complexities that arise from using transactions.
1. Possibility of deadlock between the remote execution of multiple statements 
2. Read-your-writes and monotonic write consistency need to be preserved across statements
4. Commit/abort need to be atomic across all connections

Depending on the command, we want or need parallelism through multiple connections, low latency through persistent connections, or limit the number of connections.

A transaction can be using one of the following: 

1. Cached connections to each worker (e.g. router queries)
2. Small pool of connections to each worker (e.g. DDL)
3. Connection to each shard placement (e.g. COPY)
4. Ephemeral connections for each task (e.g. real-time executor)

These can be seen as 4 different connection pool modes. 

A general observation is that we can create a mapping between shard placement and connections and each mode can be expressed with such a mapping. If a shard placement is modified over a particular connection, then all other modifications to the shard placement in that transaction should either use the same connection or throw an error. _The API for obtaining a connection should specify the shard placement and the implementation should keep track of which shard placements have been modified over which connections._

There are limited switching possibilities between the different modes within the same transaction. The cached connection could be used as a subset of the connection pool, and the connection pool can be used as a subset of connections to all placements, provided that there are no connections which modified more than 1 placement. However, there is no switching back since we need to keep all connections open until the end of the transaction.

Some commands can be performed in multiple modes. For example, if we have connections to all shard placements open due to a COPY, we could still send DML or DDL over those connections without causing problems. However, it may affect performance.

We can dynamically expand the shard ID to connection mapping to include new connections when required by the command. _The API for obtaining a connection should specify the and an intent (enum) of the caller._

**Command / mode compatibility matrix**

I'm not sure how useful this will be, but below is a mostly complete table of commands and the modes in which they could run in order to be transactional. Many of the utility commands currently run in ephemeral mode, and as such are not transactional and should either error out if used in a transaction or use the new connection API when it becomes available. Some of these functions do error out in a transaction block, but not in a function.

|                                  | cached  | small pool | all placements | ephemeral |
|----------------------------------|---------|------------|----------------|-----------|
| Router queries                   | ✓ preferred       | ✓          | ✓              |           |
| DDL                              | ?       | ✓ preferred          | ✓              |           |
| COPY                             |         |            | ✓              |           |
| Real-time executor               |         |            |                | ✓         |
| master_append_table_to_shard     | ✓       | ✓          | ✓              |           |
| master_apply_delete_command      | ✓       | ✓          | ✓              |           |
| master_create_empty_shard        | ✓       | ✓          | ✓              |           |
| master_create_worker_shards      | ✓       | ✓          | ✓              |           |
| master_copy_shard_placement      | ✓       | ✓          | ✓              |           |
| master_drop_all_shards           | ✓       | ✓          | ✓              |           |
| master_drop_sequences            | ✓       |            |                |           |
| master_modify_multiple_shards    | ✓       | ✓          | ✓              |           |
| master_update_shard_statistics   | ✓       | ✓          | ✓              | ✓         |
| cluster_create_distributed_table | ✓       |   ?         |   ?             |           |
| cluster_create_shards            | ✓       | ✓          | ✓              |           |
| MX DDL                           | ✓       |   ?         |   ?             |           |
| MX DROP                          | ✓       |   ?         |   ?             |           |

Most commands can be implemented by repeatedly obtaining a connection for a shard placement followed by executing a command, which is trivially compatible with the first 3 modes.

Parallel DDL commands are a bit trickier, since it's not obvious how the implementation would change if all shard placements are mapped to the cached connections. 

Some commands, like master_drop_sequences modify global objects on a worker and it may not make sense to send them over a shard-placement-specific connection, though it does seem technically possible.

**Real-time executor**

The real-time executor is a special case which is not obviously compatible with the other connection pooling modes than ephemeral. If there are connections open that have received writes, we could potentially reuse those connections for SELECTs to guarantee read-your-writes consistency, but this may be difficult when there are joins and subqueries.  If we have a single connection with several writes, we could alternatively make those visible in the SELECT transaction by using pg_export_snapshot on the writing transaction, and running SET TRANSACTION SNAPSHOT and setting the transaction ID within the reading transaction. For the moment, a real-time SELECT following a write will likely be unsupported. On the other hand, if the first command is a SELECT, it does not affect the transaction state since we don't currently aim to provide monotonic read consistency.

**Task-tracker executor**

The task-tracker is also a special case which is left out of scope here, since it uses an entirely separate connection manager.

**1PC vs. 2PC**

We currently use 1PC for single-worker DML transactions. We will keep doing this for performance reason. However, if a transaction requires additional connections (e.g. adds new workers or parallel connections), we switch to 2PC. We will have one commit handler that supports both these modes.

**Resource owner**

A problem with the current connection management is that in many places we carefully avoid throwing an error to make sure we eventually get to the code that closes connections. Instead we will move this into the resource owner callbacks. Some connections will be cached for the lifetime of the statement (ephemeral), transaction (small pool, all placements), or session (cached). 

**General conclusion**

There should be one API for obtaining a connection to a worker which specifies the shard placement and the intent of the caller, such that we can apply appropriate error handling, connection lifecycles, and transaction management. It was noted that it would be preferable if shard placements had a long-lived identifier. 

The initial goal of this API is to ensure that we recognise situations in which we might violate consistency or create a deadlock and error out. Later we can add more switching possibilities.

Andres plans to propose an API when he's back. Me and Jason can review and contribute to rewriting legacy code. The work is somewhat high priority since we will want to use the new API in upcoming 5.3 items like INSERT/SELECT.

**Comment by ozgune** [comment]
@aamederen @anarazel @marcocitus -- I have a question related to #709, #710, and #131. This is probably a separate issue on its own; and I wanted to check with you guys here.

When we're running ``COPY`` or DDL commands, we end up touching all shards. An improvement we could make for our existing users, who have 100K+ shards, is to error out early. We know at an early point that we won't have enough connections or max_prepared_transactions to go over 100K+ shards. (Even if we do, they may not want to run a DDL change over 1M shards with our current implementation.) Do we want to check if we have enough resources to carry out #709/#710 and fail early? Or is that a separate issue?

**Comment by anarazel** [comment]
> When we're running ``COPY`` or DDL commands, we end up touching all
> shards. An improvement we could make for our existing users, who have
> 100K+ shards, is to error out early.

Well, COPY only does so for hash partitioned tables - so a lot of connections
shouldn't really be an issue for people with 100k+ shards, as they're
certainly using append.

For DDL we're basically planning to have an API that allows to run a
list of commands efficiently over a limited number of connections.


> We know at an early point that we won't have enough connections or
> max_prepared_transactions to go over 100K+ shards. (Even if we do,
> they may not want to run a DDL change over 1M shards with our current
> implementation.) Do we want to check if we have enough resources to
> carry out #709/#710 and fail early? Or is that a separate issue?

I'd say let's get this out of the way, and then tackle the remaining
problems with too-many-connections.

**Comment by lithp** [comment]
While we're talking about this there are two other things we could worry about:
- [Propagating GUCs](https://github.com/citusdata/citus/issues/462)
- Configurable transaction isolation levels.  

Propagating GUCs was requested by a client who wanted to set `enable_seqscan` on certain queries, and I've previously found it very useful to have `application_name` set. For a SET LOCAL we forward the command to all sessions which are used by the current transaction. (All currently claimed ones, and all future ones the transaction uses). For a SET SESSION we don't take any remote action, but when a new transaction opens act as if a SET LOCAL was issued for it.

Right now we mostly provide read committed (if you ignore that we can't read-your-own-writes) but I don't think consistently. Serializable is straight out if we ever want to merge MX back into the core product but repeatable read sounds achievable if we're careful about forwarding it to the sessions we pull from the pool, much like GUCs.

--

> Parallel DDL commands are a bit trickier, since it's not obvious how the implementation would change if all shard placements are mapped to the cached connections.

> Some commands, like master_drop_sequences modify global objects on a worker and it may not make sense to send them over a shard-placement-specific connection, though it does seem technically possible.

It might be worth drawing a distinction between commands which act on nodes and global commands. In addition to a pool of `(placement, mode) -> connection` you could have a pool of `(worker,) -> connection`. Most commands would just pull a single one per worker, but parallel DDL commands would use a couple of these connections per worker.

I haven't thought about how api support aimed at preventing deadlocks between the pools would work, but as long as every caller is careful to iterate shards in in the same order it seems easy to make sure callers of the global pool don't conflict with themselves.

**Comment by anarazel** [comment]
What I think we should be doing (and what I'm prototyping) is basically to unify the following files/infrastructure to work coherently:
* connection management (and later the query handling) from multi_client_executor.h
* connection_cache.h
* multi_shard_transaction.h
* commit_protocol.h
* mx's worker_transaction.h
* multi_utility.c's  ExecuteCommandOnWorkerShards / ExecuteCommandOnShardPlacements
* router_executor.c's transaction management, including the xact participant infrastructure
* worker_data_fetch_protocol.c's ExecuteRemoteQuery()
*...

My current design, which is still in flux as I'm still prototyping, is a bit different than what any of the current ones. But it does incorporate code from various bits and pieces.  The high level picture is this:

0. Placements get an id, to be able  to efficiently and conveniently refer to them.
1. *Before* doing any remote work, using connections, and such, you do a BeginCoordinatedTransaction(). (named coordinated, since we're not really distributed...)
2. Whenever, afterwards, a connection is established that doesn't explicitly, via a flag bit, disavows itself from the transaction, it's going to automatically issue a BEGIN;
3. All connections are represented as citus defined struct (MultiConnection atm), which exposes some connection information *and* the PGconn directly.  Among that information is whether a transaction is "dirty", i.e. modified data, performed DDL
4. There's low-level connection establishment functions (GetNodeConnection(flags, host,port), GetNodeUserDatabaseConnection(flags, host,port,user,database) atm). These are supposed to, in the long run, only used by upper layer infrastructure.
5.  A higher level connection establishment layer is GetPlacementConnection(node, placementid)
6. Flags can specify whether a cached or new connection is acceptable. Unless ReturnConnection()'ed a cached connection will not be handed out again.
7. There's ExecuteCommandOnWorkers/ExecuteCommandOnShards/ExecuteCommandOnPlacements/ExecuteDDLOnPlacements/ExecuteDDLOnRelationPlacements. All of them have a flag bit that specifies whether commands have to be executed serially (parallel is default). The DDL versions automatically wrap the command in worker_apply_shard_ddl_command().  Another flag bit says whether data is modified.

Basically router executor would use GetPlacementConnection(CACHED_CONNECTION, placement). Real-Time would use GetPlacementConnection(CACHED_CONNECTION | NEW_CONNECTION). multi_utility.c would use ExecuteDDLOnRelationPlacements(), master_apply_delete_command & master_modify_multiple_shards & ... would use ExecuteCommandOnShards(), etc...

To enable that we'd basically have several hashes over the connections. One to map workers to a list of connections, one to map shards, and one to map placements. All of those need to be lists because we obviously can have multiple connections for each of those.

I'm absolutely not wedded to these function names, that's just what I ended up using in my (not yet running) prototype. Seems to mostly work
out well.

My current, unimplemented, plan is that we're automatically starting to use 2PC if more than one shardid's are in modified state. But that's optional.


One question I'm not yet entirely sure about is how to deal with marking shards as invalid in a lot of these commands. I definitely don't want to continue having these DeleteShardPlacementRow/InsertShardPlacementRow blocks all over.


@marcocitus Does that sound roughly sane? I'd not mind some input before I spend time actually making my prototype work :P.  I do expect a lot of the details to change, but I think on the higher level it should look roughly like the above.

@marcocitus @jasonmp85 Any chance you could try to finish #707 et al soon? We're going to stumble over our feets pretty badly here. And it seems a lot saner to get #707 (including phase 2) in first.

**Comment by anarazel** [comment]
> One question I'm not yet entirely sure about is how to deal with marking shards as invalid in a lot of these commands. I definitely don't want to continue having these DeleteShardPlacementRow/InsertShardPlacementRow blocks all over.

FWIW, this should probably roughly look like multi_router_executor.c's MarkRemainingInactivePlacements(), but I'm not yet sure about the exact error handling procedures. We do have slightly differing behaviour for the various commands that currently invalidate shards.

**Comment by marcocitus** [comment]
Some random comments / questions:

Will concurrent connection establishment be part of the API?

Would it make sense to have something like a List * GetConnectionsForShardModification(..) function that attempts to open connections to all placements, but if it cannot connect then the placement is marked unhealthy and not included in the list? The GetPlacementConnection function is probably only relevant for higher-level infrastructure functions and maybe real-time SELECT.

Based on that, I also wonder how badly we need placement IDs, but they're certainly nice to have.

> Before doing any remote work, using connections, and such, you do a BeginCoordinatedTransaction

Can this be done automatically using a flag that gets reset at the end of the transaction?

> There's low-level connection establishment functions (GetNodeConnection(flags, host,port), GetNodeUserDatabaseConnection(flags, host,port,user,database) atm). These are supposed to, in the long run, only used by upper layer infrastructure.

We probably still want to use these functions directly for things like shard creation.

> ExecuteDDLOnRelationPlacements

What does this do?

> ExecuteCommandOnPlacements

Random semi-relevant note: master_modify_multiple_shards serially executes commands in shard id order to avoid deadlock when there are concurrent calls that contend for the same row locks.

> All of those need to be lists because we obviously can have multiple connections for each of those.

Could we have multiple connections per placement?

> One question I'm not yet entirely sure about is how to deal with marking shards as invalid in a lot of these commands.

I think there are 2 scenarios we need to consider:
1. A commit failed in a 1PC and we want to maintain "consistency"
2. A worker is (temporarily) unreachable and we want to maintain "availability"

Scenario 1 should be straight-forward since it's already done for single-worker transactions and becomes irrelevant as soon as we switch to 2PC mode.

Scenario 2 is trickier. Unless we're pretty sure that the worker is unreachable, then we should probably roll back if a connection fails. For example, if we sent a command to a worker in the current transaction, but opening a new connection fails, rolling back is almost certainly a better option than marking the placement as inactive (e.g. we may have simply exceeded max_connections). We could even try to reuse an existing connection. However, we should do "something" if we cannot connect to a worker at all, to make sure COPY et al. remain available in case of a worker failure. 

We could introduce some node health abstraction. If a node is not healthy when we try to get a  connection for modification of a placement, then we add the placement to a list of inactive placements or error out if all placements of a shard would become unhealthy by doing so (for co-located joins, we might just aggressively try to mark all co-located placements unhealthy). Initial implementation could be something like: If there are no existing connections to a node and we fail to connect, then we mark the node as unhealthy for the duration of the transaction. If there is an existing connection and we fail to open a new one, then we error out. Eventually, NodeIsHealthy could be implemented using a separate, periodic health check to avoid sessions needlessly connecting to a dead node.

If there are any inactive placements at pre-commit time, then we should grab a global lock before marking shard states unhealthy, and after obtaining the lock, reconfirm that we're not marking all placements unhealthy.

It gets yet more complicated once we consider co-location (@onderkalaci ). We want at least one set of co-located placements to remain healthy at all times. I think we should always try to mark all co-located placements as inactive if we mark any of them, and roll back if any would violate the |healthy placements| >= 1 constraint.

**Comment by anarazel** [comment]
On 2016-08-30 05:20:51 -0700, Marco Slot wrote:
> Will concurrent connection establishment be part of the API?

Not so far, but I've bene wondering the same. Given it's performance
advantages...


> Would it make sense to have something like a List *
> GetConnectionsForShardModification(..) function that attempts to open
> connections to all placements, but if it cannot connect then the
> placement is marked unhealthy and not included in the list? The
> GetPlacementConnection function is probably only relevant for
> higher-level infrastructure functions and maybe real-time SELECT.

Hm. Maybe. But I'm not really sure for what it'd be all that interesting.


> Based on that, I also wonder how badly we need placement IDs, but they're certainly nice to have.

I don't think it's a hard "need". But it certainly makes the amount of
data to be kept around bigger, and comparisons / lookups more
cumbersome. I also already wrote a patch to change that...


> > Before doing any remote work, using connections, and such, you do a BeginCoordinatedTransaction
> 
> Can this be done automatically using a flag that gets reset at the end of the transaction?

Maybe. But I'm doubtful


> > There's low-level connection establishment functions (GetNodeConnection(flags, host,port), GetNodeUserDatabaseConnection(flags, host,port,user,database) atm). These are supposed to, in the long run, only used by upper layer infrastructure.
> 
> We probably still want to use these functions directly for things like shard creation.

Probably in some places. But I don't think shard creation should be one
of them. It's currently executed serially, with two nested loops, IIRC,
and that makes testing with higher shard counts pointlessly painful. I
don't see why we can't insert the placements locally, and then use the
normal infrastructure for DDL.


> > ExecuteDDLOnRelationPlacements
> 
> What does this do?

Execute DDL on each placement, in parallel. I.e. you pass it a list of
command strings, it wraps them in worker_apply_shard_ddl_command() and
executes them on the nodes.  Initially over #placements connections, but
I sure hope that we can reduce that to fewer connections.  Without the
automatic wrapping, you'd end up with having to have loops over
everything again.


> > ExecuteCommandOnPlacements
> 
> Random semi-relevant note: master_modify_multiple_shards serially executes commands in shard id order to avoid deadlock when there are concurrent calls that contend for the same row locks.

I doubt that actually achieves much in that specific case. Row order
isn't guaranteed to be consistent for modifications (which sucks, yes).

It generally seems good to use a consisten order.

I'm not following what executing things serially achieve here? I mean I
can see some benefit of executing things serially within the placements
of one node, but between the different shards?


> > All of those need to be lists because we obviously can have multiple connections for each of those.
> 
> Could we have multiple connections per placement?

Yes, but it should be an error to perform any modifications and such
afterwards.


> > One question I'm not yet entirely sure about is how to deal with marking shards as invalid in a lot of these commands.
> 
> I think there are 2 scenarios we need to consider:
> 1. A commit failed in a 1PC and we want to maintain "consistency"
> 2. A worker is (temporarily) unreachable and we want to maintain "availability"
> 
> Scenario 1 should be straight-forward since it's already done for single-worker transactions and becomes irrelevant as soon as we switch to 2PC mode.

I'm not so sure it's that simple. Right now we don't use transactions in
most of those cases, and simply perform commands in order. But even when
not using 2PC a lot of expensive recovery could be avoided by wrapping
nearly everything in a transaction, and only doing the COMMITs in a way
that'd require invaliding shards.

It's one thing to invalidate one placement for a simple one-row DML. But
invalidating one placement on all shards because a 1 second network
hickup is a lot worse.

So what to do really depends on caller intent :/


> Scenario 2 is trickier. Unless we're pretty sure that the worker is
> unreachable, then we should probably roll back if a connection
> fails. For example, if we sent a command to a worker in the current
> transaction, but opening a new connection fails, rolling back is
> almost certainly a better option than marking the placement as
> inactive (e.g. we may have simply exceeded max_connections). We could
> even try to reuse an existing connection. However, we should do
> "something" if we cannot connect to a worker at all, to make sure COPY
> et al. remain available in case of a worker failure.

I think it really depends on the cause of the issue, which is what makes
this hard.


> We could introduce some node health abstraction. If a node is not
> healthy when we try to get a  connection for modification of a
> placement, then we add the placement to a list of inactive placements
> or error out if all placements of a shard would become unhealthy by
> doing so (for co-located joins, we might just aggressively try to mark
> all co-located placements unhealthy).

One thing we could do is to fail the individual command, but instantly
schedule the node for more aggressive health checks.


> If there are any inactive placements at pre-commit time, then we should grab a global lock before marking shard states unhealthy, and after obtaining the lock, reconfirm that we're not marking all placements unhealthy.

That's probably going to end up being tricky from a deadlock
perspective.

**Comment by marcocitus** [comment]
> Not so far, but I've bene wondering the same. Given it's performance advantages...

And if we plan to use this API for the real-time executor we need it, otherwise we'd have a pretty significant performance regressions.

> Hm. Maybe. But I'm not really sure for what it'd be all that interesting.

Well, the question is does the transaction management API take care of shard placement health. I think it should. In that case, it would be the main API for COPY, DML, DDL, master_append_table_to_shard, master_modify_multiple_shards. Modifications always open connections to all placements except the ones they cannot connect to which are marked inactive, e.g. COPY does this using OpenCopyTransactions, the router executor does this in InitTransactionStateForTask, master_append_table_to_shard does this. DDL and master_modify_multiple_shards don't, but that's more out of simplicity than necessity. It would make sense to just have a GetConnectionsForShardModification which does this.

> Maybe. But I'm doubtful

It seems pretty similar to:
https://github.com/citusdata/citus-mx/blob/masterless/main/src/backend/distributed/transaction/worker_transaction.c#L58

>  I don't see why we can't insert the placements locally, and then use the normal infrastructure for DDL.

True.

> I doubt that actually achieves much in that specific case. Row order isn't guaranteed to be consistent for modifications (which sucks, yes).

I'm not worried about local deadlocks. PostgreSQL can clean those up. Distributed deadlocks should be avoided since we have no detection mechanism and they are very likely to occur unless we ensure order here.

> But invalidating one placement on all shards because a 1 second network hickup is a lot worse.

I don't think we need to fix that right now. We should move towards background health checks, but I think that's beyond scope.

> One thing we could do is to fail the individual command, but instantly schedule the node for more aggressive health checks.

I was thinking we could introduce a node health abstraction without actually implementing health checks yet. If we do a write while a node is unhealthy, we mark the placement on the unhealthy node as inactive.

I guess we could move to an entirely different model where we keep placements in sync using 2PC. In that case we'd also have to rethink how to recover after a node comes back. Quite possibly, we'd still want to keep track of shard placement health to know which placements have taken writes since the node became unhealthy, so making placement health work doesn't seem redundant.

> That's probably going to end up being tricky from a deadlock perspective.

Not sure why.

**Comment by anarazel** [comment]
On 2016-08-30 12:04:48 -0700, Marco Slot wrote:
> > Not so far, but I've bene wondering the same. Given it's performance advantages...
> 
> And if we plan to use this API for the real-time executor we need it, otherwise we'd have a pretty significant performance regressions.

I'm not planning to change the integration there right now, i.e. I think
it'll continue to use lower level APIs. I think that'd end up being more
invasive than I want to.  But it does make sense in the longer run.

> > Hm. Maybe. But I'm not really sure for what it'd be all that interesting.
> 
> Well, the question is does the transaction management API take care of
> shard placement health.

Well, you mentioned the router executor. I kind of understood that you
wanted to use it for select queries as well. And those don't, and
probablyu shouldn't, modify shard health. And the relevant code in the
physical planner isn't easily replacable by something like this either.

So I think we might just talked cross purposes here.


> I think it should. In that case, it would be the main API for COPY,
> DML, DDL, master_append_table_to_shard,
> master_modify_multiple_shards. Modifications always open connections
> to all placements except the ones they cannot connect to which are
> marked inactive, e.g. COPY does this using OpenCopyTransactions, the
> router executor does this in InitTransactionStateForTask,
> master_append_table_to_shard does this. DDL and
> master_modify_multiple_shards don't, but that's more out of simplicity
> than necessity. It would make sense to just have a
> GetConnectionsForShardModification which does this.

Yea, it should take care of that, I just didn't follow that that was
exactly what you meant.

> 
> > Maybe. But I'm doubtful
> 
> It seems pretty similar to:
> https://github.com/citusdata/citus-mx/blob/masterless/main/src/backend/distributed/transaction/worker_transaction.c#L58

GetWorkerTransactions() isn't really any different than
BeginCoordinatedTransaction(). I don't think it's a good idea to force
connections to be established eagerly, and such an API pretty much
forces that. There's also the issue that we'll explicitly want actions
which aren't using transactions.

**Comment by anarazel** [comment]
I really wish we could rely on 9.6's WaitEventSet api :(

**Comment by ozgune** [comment]
I read over the thread and have a few uninformed questions. No need to answer them right now - I just wanted to ask them before our v6.1 product planning sessions.

1. What are upcoming features that would benefit from these APIs? (``INSERT INTO SELECT``, shard rebalancing colocated tables, etc.)
2. I'm reading over @marcocitus's comment that [documents the compatibility matrix](https://github.com/citusdata/citus/issues/704#issuecomment-238300131). I think part of the challenge with these APIs is that they touch across many components. Is there a reduced scope that we can initially target and build from there?
3. How easy/hard is it to decouple the connection and transaction management components?
4. (less important) Do we maintain health at the worker or connection level? That is, if we fail to connect to a particular placement, chances are the worker node is unresponsive. At what level do we maintain this abstraction?

**Comment by sumedhpathak** [comment]
@anarazel Could you update the checklist in the abstract to reflect current progress? Thanks!

**Comment by anarazel** [comment]
@sumedhpathak @marcocitus Since only #1101 is open, I suggest we close this issue and use #1101 to track the remaining work?

**Comment by marcocitus** [comment]
Seems like the main infrastructure work is done and conversions are covered by #873
